### ISSUE TITLE
Migration of theme to modern responsive theme

### ISSUE THREAD

**Issue by willingc** [python/docs-community#1]
https://github.com/python/devguide/pull/679

**Comment by JulienPalard** [comment]
There's two PRs opened on python-docs-theme to modernize the current theme:

- [responsive](https://github.com/python/python-docs-theme/pull/46)
- [dark theme](https://github.com/python/python-docs-theme/pull/44)

But I do not have the CSS-fu to help here, I can only test them.

Beware if we want to switch theme, they'll have to support our language and version switchers, which is currently probably too convoluted :
- cpython's tools/templates/layout.html add placeholders in the rootrellink block.
- docsbuild switch (which knows the list of versions and languages) injects a switchers.js right before `</body>`

In the other hand, rootrellink is from sphinx, not the theme, so it could by chance just work.

**Comment by JulienPalard** [comment]
@pradyunsg, quoting you to continue in this relevant thread:

> If there's genuine interest in the modernization of the Python documentation and python-docs-theme (i.e. fairly major changes to look more like from-this-decade), I am very very interested in helping with that. :)

I think we all agree that the theme should evolve. Even from an SEO point of view (mobile-first ranking in SERPs) it's important.

As you understood, my front-end skill is absent, that's why I refrained to comment on https://github.com/python/python-docs-theme/issues/55 for example (which I followed carefully). I also avoid to merge PRs in the theme without approbations of skilled front-end developers: I can test the visual quality, but not the implementation quality.

But if you need any help understanding how the language and version switchers work, or if they need to be changed, I'll gladly help here. I recently added a github workflow on python-docs-theme to build with the version and language switchers (html downloadable as a build artifact).

**Comment by Mariatta** [comment]
I admit that I also don't have front end JS/CSS skills. I've been following all the discussions in python-docs-theme repo as well. I. just don't know how to test or how to review those changes.

**Comment by willingc** [comment]
I can take a look at the existing python-docs-theme along with Furo to determine if it makes more sense to add Python-specific stuff to Furo or vice versa. One of the nice things about Furo is it was designed with Sphinx 3.x in mind so it will bake in the advantage of updates from that, at least from a theme perspective. I've been using Furo in other projects and have been very happy with it in those.

**Comment by pradyunsg** [comment]
> language and version switchers

Yup yup! IIUC, those are the main CPython-docs-specific thing, and there's decent complexity in the build scripts due to this.

I'll take a closer look at the build scripts over this long weekend! FWIW, I'm on the PSF Slack, so if it's needed at some point, it would be possible to have a fast-than-GitHub-comments chat there.

FWIW, If there's no "strict" time pressure, I'm working on https://github.com/pradyunsg/sphinx-basic-ng/, which might actually be a better base to build upon than Furo. That + Furo's CSS which basically describes the other Sphinx theme I'm working on right now. :)

**Comment by JulienPalard** [comment]
>  I just don't know how to test

I just made this very simple: the python-docs-theme now provide a cpython built documentation [as a check artifact](https://github.com/python/python-docs-theme/pull/46/checks). It even has the version and language switchers, exactly like it's built on docs.python.org (with all the dirty tricks I'm not proud of, like editing HTML using sed).

Next step would be to temporarily (has to be defined...) host artifacts somewhere (has to be defined...) and make a bot to give the link in the PR?

**Comment by hugovk** [comment]
> Next step would be to temporarily (has to be defined...) host artifacts somewhere (has to be defined...) and make a bot to give the link in the PR?

How about including some test docs that uses the theme, and deploy to Read the Docs? We could then [enable previews from PRs](https://docs.readthedocs.io/en/stable/pull-requests.html), and it would automatically build and deploy using the theme to a test RTD site, making it easy to test the PR on desktop and mobile. 

For example, https://github.com/python/devguide/pull/679 has a preview at https://cpython-devguide--679.org.readthedocs.build/

**Comment by JulienPalard** [comment]
I have no idea how incompatible our language and version switchers are with RTD though.

**Comment by willingc** [comment]
I believe that RTD has a pretty comprehensive version switcher. We should reach out to Eric Holscher and team.

**Comment by pradyunsg** [comment]
I just flagged this discussion to @humitos, as he's in one of the members of RTD. :)

**Comment by JulienPalard** [comment]
For info, here's how the current language and version switchers in docs.python.org are working:

- [The theme places an empty div](https://github.com/python/python-docs-theme/blob/a7b271a7a92c6853b6c54797c8a1ef8e34a74566/python_docs_theme/layout.html#L19) `<div class="language_switcher_placeholder"></div>` so we can play around later.
- The theme have [the CSS style for those switchers](https://github.com/python/python-docs-theme/blob/a7b271a7a92c6853b6c54797c8a1ef8e34a74566/python_docs_theme/static/pydoctheme.css#L38).
- The docsbuild script is the only source of truth about [known versions](https://github.com/python/docsbuild-scripts/blob/328a5c2e4d7d7baad7f0d7384c69765a5c5fdd49/build_docs.py#L132) and [known languages](https://github.com/python/docsbuild-scripts/blob/328a5c2e4d7d7baad7f0d7384c69765a5c5fdd49/build_docs.py#L177).
- The docsbuild script [creates a JS file](https://github.com/python/docsbuild-scripts/blob/328a5c2e4d7d7baad7f0d7384c69765a5c5fdd49/build_docs.py#L364) with this list of languages and versions, what's needed to [create the DOM elements](https://github.com/python/docsbuild-scripts/blob/328a5c2e4d7d7baad7f0d7384c69765a5c5fdd49/templates/switchers.js#L28), and the [event handlers to do the switches](https://github.com/python/docsbuild-scripts/blob/328a5c2e4d7d7baad7f0d7384c69765a5c5fdd49/templates/switchers.js#L81).
- And now for the ugly part: the docsbuild script [injects the JS on every pages](https://github.com/python/docsbuild-scripts/blob/328a5c2e4d7d7baad7f0d7384c69765a5c5fdd49/build_docs.py#L393).

Here are some ideas on how to enhance this: https://github.com/python/docs-community/issues/4

**Comment by pradyunsg** [comment]
Okay, so... I've definitely been nerd-sniped by this. Here's an update/brain-dump on how that's going. 😅 

I've started working on a Sphinx theme, for documentation sets with the size and structure of Python's documentation. The initial working name is Lutra, which is what I'll use in this comment, but it is definitely subject to change. Here's the documentation sites I'm currently using, to draw inspirations from (additional suggestions welcome):

- [Furo] (du'oh!)
- [Ionic Framework] documentation
- [Just the Docs] for Jekyll
- [mkdocs-material] for MkDocs
- [Tailwind CSS] documentation

[furo]: https://pradyunsg.me/furo/
[ionic framework]: https://ionicframework.com/docs
[just the docs]: https://pmarsceill.github.io/just-the-docs/
[mkdocs-material]: https://squidfunk.github.io/mkdocs-material/
[tailwind css]: https://tailwindcss.com/docs

My thought process for how things _could_ work is: Lutra builds upon [`sphinx-basic-ng`](https://github.com/pradyunsg/sphinx-basic-ng/) (more on this below) and becomes the base theme for `python-docs-theme` *and* `pypa-docs-theme`, with the `python-docs-theme` being _only_ for Python's documentation. This model means we can get rid of documentation-related templates in CPython's source tree and specialise `python-docs-theme` to only be for Python, which _may_ be super useful if we want to add build-time logic (in the Sphinx extension code for `python-docs-theme`), for generating the switchers (#4) or including the top bar from https://www.python.org/ (as I'd originally proposed in https://github.com/python/python-docs-theme/issues/10).

(theme dependency tree -- status quo)

```
basic (built-in Sphinx theme)
╰─▶ default (built-in Sphinx theme)
    ╰─▶ python-docs-theme
        ╰─▶ pypa-docs-theme
```

(theme dependency tree -- what I'm currently thinking of)

```
basic (built-in Sphinx theme)
╰─▶ basic-ng
    ├─▶ [other themes]
    ╰─▶ lutra
        ├─▶ pypa-docs-theme
        ├─▶ python-docs-theme
        ╰─▶ [other themes]
```

As for sphinx-basic-ng -- that's something that will quite possibly be used by some other major themes in the Sphinx ecosystem (like sphinx-book-theme and pydata-sphinx-theme) -- I'm working closely with multiple theme authors/maintainers to figure out ways to reduce the repeated work, and have an actually thought out scaffold for this. See sphinx-basic-ng's [repository](https://github.com/pradyunsg/sphinx-basic-ng/)/[issue tracker](https://github.com/pradyunsg/sphinx-basic-ng/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc) if you're curious. I worked on it a bit last weekend and am pretty happy with where it is now, in terms of the actual implementation. It still needs a *lot* of documentation for theme authors, but slow-and-steady progress is the ongoing goal for all these things. And because video >> pictures >> words, here's a quick demo of `sphinx-basic-ng`, specifically the default behaviour of the fairly customisable scaffolding that comes with it:

https://user-images.githubusercontent.com/3275593/137479237-61c72b0a-9c75-421a-818f-35c6291ec71e.mov

**Comment by birkenfeld** [comment]
Can I just say that I love it?

**Comment by pradyunsg** [comment]
I've made some more progress, and things are at a point that it's worth sharing.

The main thing is that I've settled down on the design/aesthetic I'm going for, as well as the typography and navigation design now. I'm pretty happy with how this is working now!

https://user-images.githubusercontent.com/3275593/142778891-4eeb1202-2d8c-4e53-be77-e361a464c5dd.mov

I also ended up writing [a development workflow tool](https://github.com/pradyunsg/sphinx-theme-builder) which is a PEP 517 backend as well. So... yea. That happened. 😅 

The next things to do would be the rest of the scaffold (like the top bar, the footer, figuring out sidebar overflows etc), adding search + domain index + module index pages, and... stylizing everything that's in the metaphorical [kitchen sink](https://pradyunsg.me/furo/kitchen-sink/) for Sphinx. So... yea, baby steps.

**Comment by pradyunsg** [comment]
It's a new year! And I have made even more progress!

- There's a dark mode toggle now (auto/dark/light), and everything is stylized to respect this!

  https://user-images.githubusercontent.com/3275593/148451030-7de943a8-9a56-4e39-9c5e-28930e38cf85.mov

- I've stylized most of the content that Sphinx can output. Tables are the last remaining piece, and I still need to decide what to do with API documentation. I'm quite happy with how the anchors / `:target` styling for various things has worked out though!

  https://user-images.githubusercontent.com/3275593/148451130-af821752-d119-47b2-8eb6-229c65d0edec.mov

As you can probably tell, there's two big pieces that need to be stylized still -- the header and footer. Along with that, I also wanna get with a few bits of functionality done, before we can start testing things out with the docsbuild-scripts (eg: highlighting right sidebar entries on scroll, showing "back to top" when scrolling up, "edit this page" buttons, "was this page helpful", ability to add stuff to the top of the right sidebar etc).

I was briefly worried that this is currently written as a Sphinx 4+ theme, but it looks like the current `main` on CPython is Sphinx 4 already: https://github.com/python/cpython/blob/c9137d4b638c0699b904011cafe68895d28dd80b/Doc/requirements.txt#L6

**Comment by pradyunsg** [comment]
(oh, and I tweaked the colors to be purple out of the box, but it's like, ~10 lines of code in a downstream theme to change that -- I'd imagine that `python-docs-theme` can change that to the cyan-ish colors from the video in the comment before)

**Comment by birkenfeld** [comment]
Love it! If I was still an active Sphinx maintainer, I'd suggest that this become the new default :grin:

**Comment by ssbarnea** [comment]
@pradyunsg It looks awesome! Do we have a RTD site where we can test the theme? I like it and I would like to make that the prefered choice for most python based projects, including lots of Ansible tools my team is managing. 

It would be great if we could get an audit/review from an UX designer, especially around the accessibility areas. Making changes after adoption would be much harder so is better to do the accessibility checklist before.

I do have some previous experience with i18n/a11y but not enough to do it myself. Still, I will mention few bits

- [ ] able to use most screen sizes, portrait/landscape, mobile, different text size in browser, it should be ok even zoom at 125%
- [ ] color blindness
- [ ] display visited links differently than normal links (:hover, :visited)
- [ ] some level of branding
- [ ] Does it work well with RTD service?
- [ ] Is auto/light/dark mode preference saved using [localStorage](https://stackoverflow.com/questions/60037491/how-to-save-cookies-for-dark-light-mode-toggle) instead of cookies? Nobody likes them, or the mandatory prompts their presence require.

I do remember a nightmare related to Docker documentation which for 3+ years it was impossible to display in a narrow/split screen (even 4K/5K) as the sidebar was switching to mobile view and taking the whole page width.

**Comment by pradyunsg** [comment]
> Do we have a RTD site where we can test the theme?

No. I'll put this up once I'm ready to put things up for testing. :)

> It would be great if we could get an audit/review from an UX designer, especially around the accessibility areas.

I have spent a not-insignificant amount of time working on the accessibility aspects of Furo, so I'm not "flying blind" -- none the less, if we can get someone who has done this professionally to review these aspects, that would be great! :)

> Is auto/light/dark mode preference saved using [localStorage](https://stackoverflow.com/questions/60037491/how-to-save-cookies-for-dark-light-mode-toggle) instead of cookies?

See https://github.com/pradyunsg/furo/blob/d6b217f8bb021f137928aa30f1b41e829e87d73f/src/furo/assets/scripts/furo.js#L57, which is what this is based off.

**Comment by pradyunsg** [comment]
Hmm... I was thinking about this a bunch today, and a good question came to mind: What do people find *good* about the existing docs.python.org theme?

One potential answer that came to mind was: The collapsing sidebar. So... I've gone ahead and implemented that. I can't really think of anything else, and short of sending an email to something like python-ideas, does anyone have any suggestions for how to easily survey a bunch of folks?

**Comment by JulienPalard** [comment]
> The collapsing sidebar

I personally don't like it, but I heard of people using it.

> any suggestions for how to easily survey a bunch of folks?

Deploy the theme on docs.python.org/dev/ (I can help) and see if someone miss something? I used this strategy a lot at $DAYJOB, we called it WDD for Whine Driven Development: Push to production and raise an ear to catch if someone whines, (works best in big open spaces).

**Comment by webknjaz** [comment]
@pradyunsg add a banner/admonition to the docs like the one I've made for pip in the past, for the UX research? Our community/docs folks at Ansible do this as well putting links to surveys there.

**Comment by JulienPalard** [comment]
> add a banner

Oh yes, we do use a "deprecation notice" banner on docs.python.org too, see: https://docs.python.org/2.7/.

**Comment by pradyunsg** [comment]
> add a banner/admonition to the docs like the one I've made for pip in the past, for the UX research?

I'd prefer to instead allow adding a upvote/downvote thing, into a survey platform or whatever -- similar to the bottom of https://squidfunk.github.io/mkdocs-material/getting-started/#with-git or https://www.gatsbyjs.com/docs/quick-start.

> Oh yes, we do use a "deprecation notice" banner on docs.python.org too, see: [docs.python.org/2.7](https://docs.python.org/2.7/).

Yup! We'll treat that as an "announcement", similar to how Furo does things! :)

**Comment by encukou** [comment]
Hi! Let me play project manager here.

What's the status of this issue? With https://github.com/python/python-docs-theme/pull/46 merged, is this “done”? If not, when will it be considered done? Should individual issues be filed for the remaining work?

**Comment by pradyunsg** [comment]
> What's the status of this issue?

Still needs more work. :P

> With [python/python-docs-theme#46](https://github.com/python/python-docs-theme/pull/46) merged, is this “done”?

I don't think so. When Python documentation came up in [a Talk Python to Me podcast episode](https://talkpython.fm/episodes/transcript/354/sphinx-myst-and-python-docs-in-2022), one of the things explicitly pointed out was that docs.python.org vs something like https://tailwindcss.com/docs/ are completely different experiences -- with the latter being a lot more pleasing to look at.

> If not, when will it be considered done?

I don't think this should be considered done unless we make some progress on improving the overall site design.

> Should individual issues be filed for the remaining work?

I don't think we have a clear set of steps for what we want to do here TBH. :)

**Comment by pradyunsg** [comment]
So... that theme that I've been posting progress updates about here -- I've put the very-WIP code for it in a GitHub repository: https://github.com/pradyunsg/lutra/. The theme's own documentation is built using itself, and is available at https://pradyunsg.me/lutra/

Everything about this is very WIP but... hey, it is available publicly under MIT now! :)

**Comment by pradyunsg** [comment]
For posterity, there was some discussion about the documentation theme in the Python Documentation Community Discord (starting around https://discord.com/channels/935215565872693329/935215566334079058/975795392027447387).

---

I've gotten Lutra to a point where it's building the Python documentation without crashes and *most* elements are stylised. Here's a link to CPython's current `main` rendered with Lutra: https://pradyunsg-cpython-lutra-testing.readthedocs.io/en/latest/index.html.

I'd like to gather folks' inputs/opinions/feedback, to figure out what else needs to be changed/improved about it. To avoid hijacking this issue, I've filed https://github.com/pradyunsg/lutra/issues/22 which is where I'd request folks to post their feedback. I've also added some additional context there, for folks who are interested!

Fair warning: The theme is still WIP, so don't be surprised if you spot things that are borked/non-functional -- please let me know if/when you see that!

**Comment by AA-Turner** [comment]
Is not having an index page intentional? I think also I'd prefer if languages was spelled out as the word, I was confused by the red glyph in the navigation bar and thought it was a rendering error initially.

<details>
<summary>Screenshot</summary>

![image](https://user-images.githubusercontent.com/9087854/172261575-ca86ec06-7f37-41ed-9881-831096da2336.png)

</details>

A

**Comment by pradyunsg** [comment]
Waiting on the RTD build which will add a description for why that page is empty: https://github.com/pradyunsg/cpython/commit/f41c09bb9e7ad12af6c0528a8f6ec44c487e146f

The red glyph is mostly a placeholder reminder for me, that I have to handle language and version selectors outside of RTD builds.

**Comment by anotherk1nd** [comment]
Can this be closed now?
